ts/session/apis/open_group_api/sogsv3/sogsApiV3.ts

```diff
@@ -0,0 +1,501 @@
+         * We also need to do some custom sogs stuff like setting from which sogs conversationID this message comes from.
+         * We will need this to send new message to that user from our second device.
+         */
+        const recipient = inboxOutboxItem.recipient;
+        const contentDecoded = SignalService.Content.decode(content);
+
+        // if we already know this user's unblinded pubkey, store the blinded message we sent to that blinded recipient under
+        // the unblinded conversation instead (as we would have merge the blinded one with the )
+        const unblindedIDOrBlinded =
+          (await findCachedBlindedMatchOrItLookup(recipient, serverPubkey, sodium)) || recipient;
+
+        if (contentDecoded.dataMessage) {
+          const outboxConversationModel = await getConversationController().getOrCreateAndWait(
+            unblindedIDOrBlinded,
+            ConversationTypeEnum.PRIVATE
+          );
+          const serverConversationId = OpenGroupData.getV2OpenGroupRoomsByServerUrl(serverUrl)?.[0]
+            .conversationId;
+          if (!serverConversationId) {
+            throw new Error('serverConversationId needs to exist');
+          }
+          const msgModel = createSwarmMessageSentFromUs({
+            conversationId: unblindedIDOrBlinded,
+            messageHash: '',
+            sentAt: postedAtInMs,
+          });
+          await outboxConversationModel.setOriginConversationID(serverConversationId);
+
+          await handleOutboxMessageModel(
+            msgModel,
+            '',
+            postedAtInMs,
+            contentDecoded.dataMessage as SignalService.DataMessage,
+            outboxConversationModel
+          );
+        }
+      } else {
+        // this is an inbox message
+        const sender = inboxOutboxItem.sender;
+
+        try {
+          const match = tryMatchBlindWithStandardKey(
+            decrypted.senderUnblinded,
+            sender,
+            serverPubkey,
+            sodium
+          );
+          if (!match) {
+            throw new Error(
+              `tryMatchBlindWithStandardKey failed for blinded ${decrypted.senderUnblinded} and ${sender}`
+            );
+          }
+          // that sender just sent us its unblindedId.
+          await addCachedBlindedKey({
+            blindedId: sender,
+            realSessionId: decrypted.senderUnblinded,
+            serverPublicKey: serverPubkey,
+          });
+          await findCachedBlindedMatchOrItLookup(sender, serverPubkey, sodium);
+        } catch (e) {
+          window.log.warn('tryMatchBlindWithStandardKey could not veriyfy');
+        }
+
+        await innerHandleSwarmContentMessage(
+          builtEnvelope,
+          postedAtInMs,
+          builtEnvelope.content,
+          ''
+        );
+      }
+    } catch (e) {
+      window.log.warn('handleOutboxMessages failed with:', e.message);
+    }
+  }
+
+  const rooms = OpenGroupData.getV2OpenGroupRoomsByServerUrl(serverUrl);
```

Comment:
Am I crazy or can this check go before the for loop?

Reply:
I actually do it twice, one at the start and one at the end of the loop 
This is to take care of the case where the room was deleted while we were handling a batch of messages. So we do not create a new room with the details cached when the room was actually deleted

Reply:
Ok makes sense. Resolving.